src/opm/parser/eclipse/EclipseState/Tables/TableManager.cpp

```diff
@@ -1195,6 +1208,114 @@ namespace Opm {
+
+    void TableManager::logPVTOMonotonicityFailure(const Deck&                               deck,
+                                                  const std::size_t                         tableID,
+                                                  const std::vector<PvtoTable::FlippedFVF>& flipped_Bo) const
+    {
+        const auto& usys    = deck.getActiveUnitSystem();
+        const auto& pvtoLoc = deck.getKeyword<ParserKeywords::PVTO>(std::size_t{0}).location();
+
+        using M = UnitSystem::measure;
+        using namespace fmt::literals;
+
+        const auto nDigit = [](const std::size_t n) {
+            return 1 + static_cast<int>(std::floor(std::log10(n)));
+        };
+
+        const auto formatHeader = [&pvtoLoc](const std::size_t pvtnum)
+        {
+            return fmt::format("Non-Monotonic Oil Formation Volume Factor Detected in Keyword PVTO, PVTNUM={num}\n"
+                               "In {file} line {line}",
+                               "num"_a = pvtnum,
+                               "file"_a = pvtoLoc.filename,
+                               "line"_a = pvtoLoc.lineno);
+        };
+
+        const auto formatBoRecord =
+            [&usys](const std::size_t            numDigitsRecordID,
+                    const std::size_t            floatPrecision,
+                    const PvtoTable::FlippedFVF& flipped)
+        {
+            return fmt::format(
+                "Record {rec:{width}}: "
+                "FVF {BO_1:.{prec}f} at RS {RS_1:.{prec}f} "
+                "is not greater than "
+                "FVF {BO_0:.{prec}f} at RS {RS_0:.{prec}f}",
+                "rec"_a = flipped.i + 1,
+                "width"_a = numDigitsRecordID,
+                "prec"_a = floatPrecision,
+                "BO_1"_a = usys.from_si(M::oil_formation_volume_factor, flipped.Bo[1]),
+                "RS_1"_a = usys.from_si(M::gas_oil_ratio, flipped.Rs[1]),
+                "BO_0"_a = usys.from_si(M::oil_formation_volume_factor, flipped.Bo[0]),
+                "RS_0"_a = usys.from_si(M::gas_oil_ratio, flipped.Rs[0])
+            );
+        };
+
+        std::ostringstream prt;
+        std::ostringstream console;
+
+        {
+            const auto header = formatHeader(tableID);
+            prt     << header << '\n';
+            console << header << '\n';
+        }
+
+        // Append record to console message if either of the following conditions hold
+        //
+        //   1. Total number of flipped records does not exceed `consoleRecordLimit`.
+        //
+        //   2. Record is among `consoleRecordLimit - 1` first records.
+        //
+        // In the second case, also emit limiter message as `consoleRecordLimit`-th
+        // record.
+        //
+        // Print file gets all flipped records.
+        const auto numDigitsRecordID  = nDigit(flipped_Bo.back().i + 1);
+        const auto numRecords         = flipped_Bo.size();
+        const auto consoleRecordLimit = decltype(numRecords){4};
```

Comment:
Using `decltype(numRecords){4};` instead of `4;` reduces readability for no gain in my opinion.

Reply:
It avoid "signed vs. unsigned" warnings.

Reply:
I agree. Inside very generic library code it can pay off to be so meticulous, but I doubt its value in this context.

Reply:
The messages here got misleading order... I think applying the "const auto var = " pattern here is not necessary. Better with `std::size_t consoleRecordLimit = 4;`?

Reply:
> Better with `std::size_t consoleRecordLimit = 4;`?

Except we don't know that `std::size_t` is the correct type.  If you want to actually be correct, you need to spell it out as
```C++
const std::vector<PvtoTable::FlippedFVF>::size_type consoleRecordLimit = 4;
```
and at that point the name of the actual object (`consoleRecordLimit`) is pushed very far towards the right-hand margin.

Reply:
> Better with `std::size_t consoleRecordLimit = 4;`?

I suppose I could force all counters to `int` by writing it as
```C++
const int  numDigitsRecordID  = nDigit(flipped_Bo.back().i + 1);
const int  numRecords         = static_cast<int>(flipped_Bo.size());
const int  consoleRecordLimit = 4;
const int  floatPrecision     = 3;
const bool consoleWriteAll    = numRecords <= consoleRecordLimit;
```
but there's going to be an explicit conversion **somewhere** in that block.

Reply:
We should write code that Joe Freshman Coder would understand.

So unless necessary, we avoid decltype(). Using that tells the reader: "this is advanced stuff", and that is not true here. Also, std::size_t is *a* correct type to use here, the full size_type specification is over-engineering it, reduces readability, and, again, tells the reader "this is advanced".

About forcing counters to `int`: I used to advocate that, but I have for a long time acknowledged that battle is lost, so whenever we need to compare integer types, follow the path of least resistance. I do not know from the code snippet what parts are unsigned or not. Outside very generic utility libraries I do not consider that a benefit, but a drawback. It is fine to use auto, but use a specific type when necessary.

Would not the simplest solution avoiding a warning be `const auto consoleRecordLimit = 4u;`?

Reply:
> the full `size_type` specification is over-engineering it, reduces readability, and, again, tells the reader "this is advanced".

Which is why `auto` is so much better for this.  It tells the reader, "I don't really care what the particular type of this object is, just make it the same as that initializer over there".

In any case, I've pushed an update that forces the counters to `std::size_t`.  This is preferable to using `auto = 4u` because that will generate `unsigned int` which is definitely wrong on many 64 bit platforms.